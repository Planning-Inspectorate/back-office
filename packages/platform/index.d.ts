import './config/dotenv';
import './http/ttl';
import './util/CryptoUtils';
import './util/date';
import './util/filter';
import './util/url';
import './validators/date';
import './validators/postcode';

declare module '@pins/platform' {
	export * from './config/dotenv';
	export * from './http/ttl';
	export * from './util/CryptoUtils';
	export * from './util/date';
	export * from './util/filter';
	export * from './util/url';
	export * from './validators/date';
	export * from './validators/postcode';
}
