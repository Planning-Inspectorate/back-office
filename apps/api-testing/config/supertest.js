// @ts-nocheck
import supertest from 'supertest';
import config from './config.cjs';
import { Logger } from '../utils/logger.js';

const logger = new Logger();
logger.log('green', '************* Tests starting ****************');
logger.log('cyan', `Running tests against: ${config.baseUrl}`);
logger.log('green', '*********************************************');

const request = supertest(config.baseUrl);

export default request;
