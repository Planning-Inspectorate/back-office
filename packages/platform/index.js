export * from './http/hooks.js';
export * from './testing/assets/assets.js';
export * from './testing/fake.js';
export * from './testing/html-parser.js';
export * from './utils/date.js';
export * from './utils/dotenv.js';
export * from './utils/filter.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
