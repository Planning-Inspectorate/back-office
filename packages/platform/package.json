{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=16.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.0",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.1",
    "@types/lodash-es": "^4.17.6",
    "type-fest": "^2.12.2"
  }
}
