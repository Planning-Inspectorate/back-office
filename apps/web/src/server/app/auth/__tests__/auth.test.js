import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import {
	ConfidentialClientApplication,
	createAccountInfo
} from '../../../../../testing/app/app.js';
import { createTestEnvironment } from '../../../../../testing/index.js';

const { app, teardown } = createTestEnvironment({ authenticated: false });
const request = supertest(app);
const azureMsalMock = ConfidentialClientApplication.getMock();

describe('auth', () => {
	afterEach(teardown);

	describe('authentication', () => {
		it('should sign in an unauthenticated user via azure SSO', async () => {
			const response = await request.get('/appeals-service/validation').redirects(1);
			const client = getConfidentialClientApplication();

			// Assert azure msal client was instantiated correctly

			expect(client?.configuration.auth).toEqual({
				authority: 'https://login.microsoftonline.com/auth_tenant_id',
				clientId: 'auth_client_id',
				clientSecret: 'auth_client_secret'
			});

			// Assert azure msal client's `getAuthCodeUrl` was invoked correctly

			const [authOptions] = /** @type {import('@azure/msal-node').AuthorizationUrlRequest[]} */ (
				client.getAuthCodeUrl.mock.lastCall
			);

			expect(authOptions.authority).toEqual('https://login.microsoftonline.com/auth_tenant_id');
			expect(authOptions.redirectUri).toEqual(expect.stringMatching('^https?://.*/auth/redirect'));
			expect(authOptions.scopes).toEqual(['user.read']);
			expect(authOptions.nonce).toBeTruthy();

			expect(response.get('Location')).toEqual('/test/azure-msal/signin');
			// Invoke the redirect from azure's SSO

			const redirect = await request.get('/auth/redirect?code=msal_code');
			const [tokenOptions] = /** @type {import('@azure/msal-node').AuthorizationCodeRequest[]} */ (
				client.acquireTokenByCode.mock.lastCall
			);

			expect(tokenOptions.authority).toEqual('https://login.microsoftonline.com/auth_tenant_id');
			expect(tokenOptions.redirectUri).toEqual(expect.stringMatching('^https?://.*/auth/redirect'));
			expect(tokenOptions.scopes).toEqual(['user.read']);
			expect(tokenOptions.code).toEqual('msal_code');

			expect(redirect.get('Location')).toEqual('/appeals-service/validation');
		});

		it('should redirect to an error page when the redirect from MSAL is incomplete', async () => {
			const response = await request.get('/auth/redirect').redirects(1);
			const element = parseHtml(response.text);

			expect(element.querySelector('h1')?.innerHTML).toEqual(
				'Sorry, there is a problem with the service'
			);
		});

		it('should redirect to an error page when the nonce is absent from the acquired token', async () => {
			await request.get('/').redirects(1);

			const client = getConfidentialClientApplication();

			client.acquireTokenByCode.mockImplementationOnce(() =>
				Promise.resolve(
					/** @type {import('@azure/msal-node').AuthenticationResult} */ ({
						account: client.account,
						// omit expected `nonce` value from authentication result
						idTokenClaims: {}
					})
				)
			);

			const response = await request.get('/auth/redirect?code=msal_code').redirects(1);
			const element = parseHtml(response.text);

			expect(element.querySelector('h1')?.innerHTML).toEqual(
				'Sorry, there is a problem with your login'
			);
		});

		it('should redirect to an error page when the nonce in the acquired token does not match', async () => {
			await request.get('/').redirects(1);

			const client = getConfidentialClientApplication();

			client.acquireTokenByCode.mockImplementationOnce(() =>
				Promise.resolve(
					/** @type {import('@azure/msal-node').AuthenticationResult} */ ({
						account: client.account,
						// omit expected `nonce` value from authentication result
						idTokenClaims: {
							nonce: '*'
						}
					})
				)
			);

			const response = await request.get('/auth/redirect?code=msal_code').redirects(1);
			const element = parseHtml(response.text);

			expect(element.querySelector('h1')?.innerHTML).toEqual(
				'Sorry, there is a problem with your login'
			);
		});

		it('should redirect to an error page when fetching the auth url fails', async () => {
			getConfidentialClientApplication().getAuthCodeUrl.mockImplementationOnce(() =>
				Promise.reject(new Error('*'))
			);

			const response = await request.get(`/`).redirects(1);
			const element = parseHtml(response.text);

			expect(element.querySelector('h1')?.innerHTML).toEqual(
				'Sorry, there is a problem with the service'
			);
		});

		it('should redirect to an error page when acquiring the access token fails', async () => {
			await request.get('/auth/signin');

			getConfidentialClientApplication().acquireTokenByCode.mockImplementationOnce(() =>
				Promise.reject(new Error('*'))
			);

			const response = await request.get('/auth/redirect?code=msal_code').redirects(1);
			const element = parseHtml(response.text);

			expect(element.querySelector('h1')?.innerHTML).toEqual(
				'Sorry, there is a problem with the service'
			);
		});

		it('should silently reacquire a token on each route navigation', async () => {
			await signinWithGroups(['appeals_validation_officer']);
			await request.get('/');

			const client = getConfidentialClientApplication();

			expect(client.acquireTokenSilent).toHaveBeenCalledWith({
				account: client.account,
				scopes: ['user.read']
			});
		});

		it('should clear any pending authentication data if the MSAL redirect is not subsequently invoked', async () => {
			// start the MSAL authentication flow
			await request.get('/auth/signin');
			// trigger an intermediary request while awaiting MSAL redirect
			await request.get('/');

			// complete MSAL redirect after authentication data has been cleared
			const response = await request.get('/auth/redirect?code=msal_code').redirects(1);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should display a logout link in the header when the user is authenticated', async () => {
			// visit an unauthenticated route
			let response = await request.get(`/unauthenticated`);
			let element = parseHtml(response.text, { rootElement: '.govuk-header' });

			expect(element.innerHTML).toMatchSnapshot();

			await signinWithGroups(['appeals_validation_officer']);
			response = await request.get(`/unauthenticated`);
			element = parseHtml(response.text, { rootElement: '.govuk-header' });

			expect(element.innerHTML).toMatchSnapshot();
		});

		it('should destroy the msal token cache and session upon logging out', async () => {
			await signinWithGroups(['appeals_validation_officer']);
			await request.get('/auth/signout');

			// access an authenticated route to determine if we're signed out
			const response = await request.get('/validation').redirects(1);

			expect(response.get('Location')).toEqual('/test/azure-msal/signin');
		});
	});

	describe('authorization', () => {
		it('should display the services page when a user belongs to more than one group', async () => {
			await signinWithGroups(['appeals_validation_officer', 'applications_case_team']);

			const response = await request.get('/');
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();
		});

		describe('/appeals-service/appeals-list', () => {
			beforeEach(() => {
				nock('http://test/').get('/appeals-service/appeals-list').reply(200, []);
			});

			it('should deny access to the domain if the user does not have permission', async () => {
				await signinWithGroups(['appeals_inspector', 'appeals_case_officer']);

				const response = await request.get('/appeals-service/appeals-list').redirects(1);
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toEqual(
					'Sorry, there is a problem with your login'
				);
			});

			it('should permit access to the domain if the user has permission', async () => {
				await signinWithGroups(['appeals_validation_officer']);

				const response = await request.get('/appeals-service/appeals-list');
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toEqual('Appeals');
			});

			it('should redirect to the national-list page from the root path', async () => {
				await signinWithGroups(['appeals_validation_officer']);

				const response = await request.get('/').redirects(1);
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toEqual('Appeals');
			});
		});

		describe('/applications-service/case-admin-officer', () => {
			beforeEach(() => {
				nock('http://test/').get('/applications/case-admin-officer').reply(200, []);
			});

			it('should deny access to the domain if the user does not have permission', async () => {
				await signinWithGroups(['applications_case_team', 'applications_case_team']);

				const response = await request.get('/applications-service/inspector').redirects(1);
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toEqual(
					'Sorry, there is a problem with your login'
				);
			});

			it('should permit access to the domain if the user has permission', async () => {
				await signinWithGroups(['applications_case_admin_officer']);

				const response = await request.get('/applications-service/case-admin-officer');
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toContain('Applications');
			});

			it('should redirect to the case admin officer page from the root path', async () => {
				await signinWithGroups(['applications_case_admin_officer']);

				const response = await request.get('/').redirects(1);
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toContain('Applications');
			});
		});

		describe('/applications-service/case-team', () => {
			beforeEach(() => {
				nock('http://test/').get('/applications/case-team').reply(200, []);
			});

			it('should deny access to the domain if the user does not have permission', async () => {
				await signinWithGroups(['applications_case_admin_officer', 'applications_inspector']);

				const response = await request.get('/applications-service/case-team').redirects(1);
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toEqual(
					'Sorry, there is a problem with your login'
				);
			});

			it('should permit access to the domain if the user has permission', async () => {
				await signinWithGroups(['applications_case_team']);

				const response = await request.get('/applications-service/case-team');
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toContain('Applications');
			});

			it('should redirect to the case officer page from the root path', async () => {
				await signinWithGroups(['applications_case_team']);

				const response = await request.get('/').redirects(1);
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toContain('Applications');
			});
		});

		describe('/applications-service/inspector', () => {
			beforeEach(() => {
				nock('http://test/').get('/applications/inspector').reply(200, []);
			});

			it('should deny access to the domain if the user does not have permission', async () => {
				await signinWithGroups(['applications_case_admin_officer', 'applications_case_team']);

				const response = await request.get('/applications-service/inspector').redirects(1);
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toEqual(
					'Sorry, there is a problem with your login'
				);
			});

			it('should permit access to the domain if the user has permission', async () => {
				await signinWithGroups(['applications_inspector']);

				const response = await request.get('/applications-service/inspector');
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toContain('Applications');
			});

			it('should redirect to the inspector page from the root path', async () => {
				await signinWithGroups(['applications_inspector']);

				const response = await request.get('/').redirects(1);
				const element = parseHtml(response.text);

				expect(element.querySelector('h1')?.innerHTML).toContain('Applications');
			});
		});
	});
});

/** @returns {ConfidentialClientApplication} */
function getConfidentialClientApplication() {
	return /** @type {ConfidentialClientApplication} */ (
		azureMsalMock.confidentialClientApplications.last
	);
}

/** @typedef {'appeals_validation_officer' | 'appeals_case_officer' | 'appeals_inspector'} AppealGroupId  */
/** @typedef {'applications_case_admin_officer' | 'applications_case_team' | 'applications_inspector'} ApplicationsGroupId  */

/**
 * @param {Array<AppealGroupId | ApplicationsGroupId>} groups
 * @returns {Promise<void>}
 */
async function signinWithGroups(groups) {
	getConfidentialClientApplication().account = createAccountInfo({ groups });

	await request.get('/').redirects(1);
	await request.get(`/auth/redirect?code=msal_code`);
}
