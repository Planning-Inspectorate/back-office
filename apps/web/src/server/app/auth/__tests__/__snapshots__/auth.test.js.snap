// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth authentication should clear any pending authentication data if the MSAL redirect is not subsequently invoked 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-grid-row\\">
        <div class=\\"govuk-grid-column-two-thirds\\">
            <h1 class=\\"govuk-heading-xl\\">Sorry, there is a problem with the service</h1>
            <p class=\\"govuk-body\\">Try again later.</p>
            <p class=\\"govuk-body\\"><a href=\\"https://has-appeal.herokuapp.com/help/contact\\" class=\\"govuk-link\\">Contact the Planning Inspectorate Customer Support</a> if
                the problem persists.</p>
            <hr>
            <h1 class=\\"govuk-heading-m\\" style=\\"color: #d4351c;\\">Error (only seen in dev mode):</h1>
            <div class=\\"govuk-grid-row\\">
                <div class=\\"govuk-grid-column-full\\">
                    <div>Error: Authentication does not exist.</div>}<pre style=\\"border: 1px solid #ddd; overflow: scroll\\"> Error: Authentication does not exist.
    at Module.getAuthenticationData (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/apps/web/src/server/app/auth/auth-session.service.js:32:8)
    at completeMsalAuthentication (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/apps/web/src/server/app/auth/auth.controller.js:57:55)
    at Layer.handleRequest (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/node_modules/router/lib/layer.js:101:15)
    at next (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/node_modules/router/lib/route.js:142:13)
    at assertIsUnauthenticated (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/apps/web/src/server/app/auth/auth.guards.js:56:3)
    at Layer.handleRequest (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/node_modules/router/lib/layer.js:101:15)
    at next (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/node_modules/router/lib/route.js:142:13)
    at Route.dispatch (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/node_modules/router/lib/route.js:107:3)
    at handle (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/node_modules/router/index.js:421:11)
    at Layer.handleRequest (/Users/francescobrenci/Documents/DEVELOPMENT/back-office/node_modules/router/lib/layer.js:101:15)</pre>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`auth authentication should display a logout link in the header when the user is authenticated 1`] = `
"<div class=\\"govuk-header__content\\">
    <nav aria-label=\\"Menu\\" class=\\"govuk-header__navigation pins-header__navigation--align-right\\">
        <button type=\\"button\\" class=\\"govuk-header__menu-button govuk-js-header-toggle\\"
        aria-controls=\\"navigation\\" aria-label=\\"Show or hide menu\\">Menu</button>
        <ul id=\\"navigation\\" class=\\"govuk-header__navigation-list\\"></ul>
    </nav>
</div>"
`;

exports[`auth authentication should display a logout link in the header when the user is authenticated 2`] = `
"<div class=\\"govuk-header__content\\">
    <nav aria-label=\\"Menu\\" class=\\"govuk-header__navigation pins-header__navigation--align-right\\">
        <button type=\\"button\\" class=\\"govuk-header__menu-button govuk-js-header-toggle\\"
        aria-controls=\\"navigation\\" aria-label=\\"Show or hide menu\\">Menu</button>
        <ul id=\\"navigation\\" class=\\"govuk-header__navigation-list\\">
            <li class=\\"govuk-header__navigation-item\\"><a class=\\"govuk-header__link\\" href=\\"https://login.microsoftonline.com/common/oauth2/v2.0/logout\\"> Sign out</a>
            </li>
        </ul>
    </nav>
</div>"
`;

exports[`auth authorization should display the services page when a user belongs to more than one group 1`] = `
"<main class=\\"govuk-main-wrapper \\" id=\\"main-content\\" role=\\"main\\">
    <div class=\\"govuk-width-container govuk-width-container\\">
        <div class=\\"govuk-grid-row\\">
            <div class=\\"govuk-grid-column-one-half\\">
                <h2 class=\\"govuk-heading-m\\">Applications</h2>
                <ul class=\\"govuk-list\\">
                    <li><a href=\\"/applications-service/case-officer\\" class=\\"govuk-link\\">Case officer</a>
                    </li>
                </ul>
            </div>
            <div class=\\"govuk-grid-column-one-half\\">
                <h2 class=\\"govuk-heading-m\\">Appeals</h2>
                <ul class=\\"govuk-list\\">
                    <li><a href=\\"/appeals-service/validation\\" class=\\"govuk-link\\">Validation officer</a>
                    </li>
                </ul>
            </div>
        </div>
    </div>
</main>"
`;
